File: native/app/contribution_receipt_content.go
package app

import (
	"errors"
	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/sites"
)

func contributionReceiptManifestPolicy(m core.Manifest) bool {
	if m.Kind != "site-contribution-receipt" || m.PublicKey != nil || len(m.Keys) < 1 || len(m.Keys) > 2 || m.Expires-m.Created > sites.ContributionReceiptLifetimeMS {
		return false
	}
	total := 0
	for _, chunk := range m.Chunks {
		total += chunk.Size
	}
	if total > sites.ContributionReceiptBytes+128 {
		return false
	}
	for _, key := range m.Keys {
		if key.Reader == m.Author.ID {
			return true
		}
	}
	return false
}
func inspectContributionReceipt(bundle core.Bundle, identity *core.Identity) (map[string]any, error) {
	if bundle.Manifest.Kind != "site-contribution-receipt" {
		return nil, nil
	}
	if !contributionReceiptManifestPolicy(bundle.Manifest) {
		return nil, errors.New("recibos exigem um envelope privado limitado")
	}
	readable := false
	for _, key := range bundle.Manifest.Keys {
		if identity != nil && key.Reader == identity.Public.ID {
			readable = true
		}
	}
	if !readable {
		return nil, nil
	}
	plain, err := core.DecryptBundle(bundle, identity)
	if err != nil {
		return nil, err
	}
	return sites.MatchContributionReceiptEnvelope(bundle, plain)
}

File: apps/node/src/contribution-receipt-content.ts
import {
  decryptBundle,
  type Bundle,
  type Identity,
} from "../../../packages/core/src/index";
import { nodeCertificateCrypto } from "../../../packages/core/src/certificate-crypto";
import { contributionReceiptManifestPolicy } from "../../../packages/content/src/receipt-content";
import { createContributionReceiptProtocol } from "../../../packages/sites/src/contribution-receipt";
const protocol = createContributionReceiptProtocol(nodeCertificateCrypto);
export function inspectContributionReceipt(
  bundle: Bundle,
  identity?: Identity,
) {
  if (bundle.manifest.kind !== "site-contribution-receipt") return null;
  if (!contributionReceiptManifestPolicy(bundle.manifest))
    throw Error("Recibos exigem um envelope privado limitado");
  if (
    !identity ||
    !bundle.manifest.keys.some((k) => k.reader === identity.public.id)
  )
    return null;
  return protocol.matchEnvelope(bundle, decryptBundle(bundle, identity));
}
